lab1/myswitch_stp.py

```diff
@@ -0,0 +1,247 @@
+import sys
+import time
+from datetime import datetime
+
+import SpanningTreeMessage as STM
+from switchyard.lib.userlib import *
+
+
+class ForwardingTable:
+    def __init__(self, size=5):
+        self.array = [None for i in range(size)]
+        self.pos = 0
+        self.size = size
+        self.map = {}
+        return
+
+    def __iter__(self):
+        return iter(self.map)
+
+    def __getitem__(self, item):
+        return self.map[item]
+
+    def update(self, addr, port):
+        """ if addr is already in map, we set it again
+        without changing anything else, (as the port may
+        have changed).
+        If addr not in map, we update ForwardingTable in
+        a fifo manner, e.g. if we have a vacant spot, we
+        will just add the new addr/port pair; otherwise,
+        if we do not have a vacancy, we'll need to evict
+        the oldest item from the ForwardingTable, then add
+        the new item
+        """
+        if addr in self.map:
+            self.map[addr] = port
+            return
+
+        self.map[addr] = port
+        if self.array[self.pos] == None:
+            self.array[self.pos] = addr
+        else:
+            evict = self.array[self.pos]
+            del self.map[evict]
+            self.array[self.pos] = addr
+
+        self.pos = (self.pos + 1) % self.size
+
+        return
+
+
+class SpanningTreeContext:
+    def __init__(self, my_id):
+        """ switches assume they're the root node upon startup
+        """
+        self.my_id = EthAddr(my_id)
+        self.root_id = EthAddr(my_id)
+        self.hops_from_root = 0
+        self.time_last_spm_tx = None
+        """ Non-root mode properties
+        """
+        # rx interface from where the STM of the perceived root came
+        self.root_interface = None
+        # ID of the switch connected to the root_interface
+        self.root_switch_id = None
```

Comment:
root_switch_id and root_id refer to the same concept; root_id is never updated once it's initialized to EthAddr(my_id)

Reply:
See my reply above